webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/reports/execution/ReportsExecutionRepository.scala

```diff
@@ -133,67 +135,44 @@ class CachedReportsExecutionRepository(
-            case _ =>
-                Left(AgentRunWithNodeConfig(x.agentRunId, x.nodeConfigVersion.map(c => (c, None)), x.isCompleted, x.insertionId))
-          }
+    (for {
+      unprocessedRuns <- readBackend.getUnprocessedRuns
+      // first evolution, get same behaviour than before, and returns only the last run per node
+      // ignore those without a nodeConfigId
+      lastRunByNode = unprocessedRuns.filter(_.nodeConfigVersion.isDefined).groupBy(_.agentRunId.nodeId).map {
+        case (nodeid, seq) => (nodeid, seq.sortBy(_.agentRunId.date).last)
       }
-    }
-    val n2 = System.currentTimeMillis
-    TimingDebugLogger.trace(s"CachedReportsExecutionRepository: update execution in ${n2 - n1}ms")
-
-    //now, get back the one in left, and query for node config
-    val nodeAndConfigs = runWithConfigs.collect { case Left(AgentRunWithNodeConfig(id, Some((config, None)), _, _)) => NodeAndConfigId(id.nodeId, config)}
-
-    val results = (findConfigs.getExpectedReports(nodeAndConfigs.toSet) match {
-      case eb: EmptyBox =>
-        val e = eb ?~! s"Error when trying to find node configuration matching new runs for node/configId: ${nodeAndConfigs.map(x=> s"${x.nodeId.value}/${x.version.value}").mkString(", ")}"
-        logger.error(e.messageChain)
-        //return the whole list of runs unmodified
-        runWithConfigs.map {
-          case Left(x)  => (x.agentRunId.nodeId, Some(x))
-          case Right(x) => (x.agentRunId.nodeId, Some(x))
-        }
-      case Full(map) =>
-        runWithConfigs.map {
-          case Left(x)  =>
-            val configVersion = x.nodeConfigVersion.map(c => c.copy(_2 = map.get(NodeAndConfigId(x.agentRunId.nodeId, c._1)).flatten))
-            (x.agentRunId.nodeId, Some(x.copy(nodeConfigVersion = configVersion)))
-          case Right(x) => (x.agentRunId.nodeId, Some(x))
-        }
-    }).toMap
-    val n3 = System.currentTimeMillis
-    TimingDebugLogger.trace(s"CachedReportsExecutionRepository: compte result of update execution in ${n3 - n2}ms")
-
-    cache = cache ++ results
-    runs
+      // by construct, we do have a nodeConfigId
+      agentsRuns =   lastRunByNode.map(x => (x._1, NodeAndConfigId(x._1, x._2.nodeConfigVersion.get)))
+
+      expectedReports <- findConfigs.getExpectedReports(agentsRuns.values.toSet).toIO
+
+      runs = agentsRuns.map { case (nodeId, nodeAndConfigId) =>
+        (nodeId,
+          Some(AgentRunWithNodeConfig(
+               AgentRunId(nodeId, lastRunByNode(nodeId).agentRunId.date)
+             , expectedReports.get(nodeAndConfigId).map(optionalExpectedReport => (nodeAndConfigId.version, optionalExpectedReport))
+             , true
+             , lastRunByNode(nodeId).insertionId)))}
+      // and finally mark them read. It's so much easier to do it now than to carry all data all the way long
+      _  <- writeBackend.setComplianceComputationDate(unprocessedRuns.toList)
+    } yield {
```

Comment:
I think the cache doesn't have access to the getExpectedReports

Reply:
indeed, the backend is simply jdbc, and doesn't have access to other services. I'd rather leave it like that, as next iteration will change the getExpectedReports part significantly, with a cache

Reply:
I would prefer to have a totally fine, perfect by itself iteration, because we never know when the following one can be - especially true in current context :)

Reply:
ok but it will be messy